desc/objectives/_free_boundary.py

```diff
@@ -0,0 +1,882 @@
+                "Could not built fft interpolator, switching to dft method which is"
+                " much slower. Reason: " + str(e)
+            )
+            interpolator = DFTInterpolator(eval_grid, src_grid, self._s, self._q)
+
+        edge_pres = np.max(np.abs(eq.compute("p", grid=eval_grid)["p"]))
+        warnif(
+            (edge_pres > 1e-6) and not self._sheet_current,
+            UserWarning,
+            f"Boundary pressure is nonzero (max {edge_pres} Pa), "
+            + "a sheet current should be included.",
+        )
+
+        self._eq_data_keys = [
+            "K_vc",
+            "B",
+            "|B|^2",
+            "R",
+            "zeta",
+            "Z",
+            "e^rho",
+            "n_rho",
+            "|e_theta x e_zeta|",
+            "p",
+        ]
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        src_profiles = get_profiles(self._eq_data_keys, obj=eq, grid=src_grid)
+        src_transforms = get_transforms(self._eq_data_keys, obj=eq, grid=src_grid)
+        eval_profiles = get_profiles(self._eq_data_keys, obj=eq, grid=eval_grid)
+        eval_transforms = get_transforms(self._eq_data_keys, obj=eq, grid=eval_grid)
+
+        neq = 3 if self._sheet_current else 2  # number of equations we're using
+
+        self._constants = {
+            "eval_transforms": eval_transforms,
+            "eval_profiles": eval_profiles,
+            "src_transforms": src_transforms,
+            "src_profiles": src_profiles,
+            "interpolator": interpolator,
+            "ext_field": self._ext_field,
+            "quad_weights": np.sqrt(np.tile(eval_transforms["grid"].weights, neq)),
+        }
+
+        if self._sheet_current:
+            sheet_current = self.things[1]
+            assert (
+                (sheet_current.M == eq.surface.M)
+                and (sheet_current.N == eq.surface.N)
+                and (sheet_current.NFP == eq.surface.NFP)
+                and (sheet_current.sym == eq.surface.sym)
+            ), "sheet current must have same resolution as eq.surface"
+            self._sheet_data_keys = ["K"]
+            sheet_eval_transforms = get_transforms(
+                self._sheet_data_keys, obj=sheet_current, grid=eval_grid
+            )
+            sheet_src_transforms = get_transforms(
+                self._sheet_data_keys, obj=sheet_current, grid=src_grid
+            )
+            self._constants["sheet_eval_transforms"] = sheet_eval_transforms
+            self._constants["sheet_src_transforms"] = sheet_src_transforms
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        self._dim_f = neq * eval_grid.num_nodes
+
+        if self._normalize:
+            scales = compute_scaling_factors(eq)
+            self._normalization = scales["B"] * scales["R0"] * scales["a"]
```

Comment:
same comment as above, is this normalization actually the correct one for the other errors (in `K` and in `Bsq`)?